src/transaction/flashback.c

```diff
@@ -0,0 +1,416 @@
+#include "oid.h"
+#include "storage_common.h"
+
+static int
+flashback_create_summary_entry (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_ENTRY ** summary)
+{
+  *summary = (FLASHBACK_SUMMARY_ENTRY *) db_private_alloc (thread_p, sizeof (FLASHBACK_SUMMARY_ENTRY));
+  if (*summary == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (FLASHBACK_SUMMARY_ENTRY));
+      return ER_OUT_OF_VIRTUAL_MEMORY;
+    }
+
+  (*summary)->trid = NULL_TRANID;
+  (*summary)->user[0] = '\0';
+  (*summary)->start_time = -1;
+  (*summary)->end_time = -1;
+  (*summary)->num_insert = 0;
+  (*summary)->num_update = 0;
+  (*summary)->num_delete = 0;
+  (*summary)->start_lsa = LSA_INITIALIZER;
+  (*summary)->end_lsa = LSA_INITIALIZER;
+  (*summary)->num_class = 0;
+  (*summary)->classlist[0] = OID_INITIALIZER;
+
+  return NO_ERROR;
+}
+
+static void
+flashback_drop_summary_entry (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_ENTRY ** summary)
+{
+  assert (*summary != NULL);
+
+  db_private_free_and_init (thread_p, *summary);
+}
+
+/*
+ * flashback_fill_dml_summary - fill the dml information into summary entry
+ *
+ * thread_p (in)             : thread entry
+ * summary_entry (in/out)    : flashback summary entry
+ * classoid (in)             : OID of DML target class
+ * rec_type(in)              : record type that can be classified as DML types
+ */
+
+static void
+flashback_fill_dml_summary (FLASHBACK_SUMMARY_ENTRY * summary_entry, OID classoid, SUPPLEMENT_REC_TYPE rec_type)
+{
+  assert (summary_entry != NULL);
+
+  bool found = false;
+  int i = 0;
+
+  /* fill in the summary_entry info entry
+   * 1. number of class and put the classoid into the classlist
+   * 2. number of DML operation */
+
+  for (i = 0; i < summary_entry->num_class; i++)
+    {
+      if (OID_EQ (&summary_entry->classlist[i], &classoid))
+	{
+	  found = true;
+	  break;
+	}
+    }
+
+  if (!found)
+    {
+      summary_entry->num_class += 1;
+      COPY_OID (&summary_entry->classlist[i], &classoid);
+    }
+
+  switch (rec_type)
+    {
+    case LOG_SUPPLEMENT_INSERT:
+    case LOG_SUPPLEMENT_TRIGGER_INSERT:
+      summary_entry->num_insert++;
+      break;
+    case LOG_SUPPLEMENT_UPDATE:
+    case LOG_SUPPLEMENT_TRIGGER_UPDATE:
+      summary_entry->num_update++;
+      break;
+    case LOG_SUPPLEMENT_DELETE:
+    case LOG_SUPPLEMENT_TRIGGER_DELETE:
+      summary_entry->num_delete++;
+      break;
+    default:
+      assert (false);
+      break;
+    }
+
+  return;
+}
+
+/*
+ * flashback_cleanup - dealloc memory of summary entries iterating summary list
+ *
+ * thread_p (in)         : thread entry
+ * context (in/out)      : flashback summary context that contains summary list
+ */
+
+void
+flashback_cleanup (THREAD_ENTRY * thread_p, FLASHBACK_SUMMARY_CONTEXT * context)
+{
+  /* *INDENT-OFF* */
+  for (auto iter:context->summary_list)
+    {
+      if (iter.second != NULL)
+        {
+	  db_private_free_and_init (thread_p, iter.second);
```

Comment:
How about calling the flashback_drop_summary_entry function?

Reply:
Ok, it will be more readable